spacy_transformers/util.py

```diff
@@ -1,426 +1,214 @@
-        tokens:   [CLS] the dog is hairy . [SEP]
-        type_ids:   0   0   0   0  0     0   0
-    """
-    if length2:
-        return [0] * length1 + [0] + [0] + [1] * length2 + [1]
-    else:
-        return [0] * length1 + [0] + [0]
-
-
-def get_xlnet_segment_ids(length1: int, length2: int) -> List[int]:
-    """Get an array of segment IDs in XLNet's format, for an input with one or
-    two segments (set length2=0 for one segment). The lengths should be just the
-    wordpiece lengths, not including the SEP and CLS tokens.
-
-    According to the XLNet code classifer_utils.py module, the convention is:
-
-    (a) For sequence pairs:
-        tokens:    is this jack ##son ##ville ? <sep> no it is not . <sep> <cls>
-        type_ids:   0   0  0    0      0      0  0    1   1  1  1  1   1   2
-
-    (b) For single sequences:
-        tokens:   the dog is hairy . <sep> <cls>
-        type_ids:   0   0 0   0    0   0   2
-    """
-    if length2:
-        return [0] * length1 + [0] + [1] * length2 + [1, 2]
-    else:
-        return [0] * length1 + [0, 2]
-
-
-def get_xlm_segment_ids(length1: int, length2: int) -> List[int]:
-    """Get an array of segment IDs in XLNet's format, for an input with one or
-    two segments (set length2=0 for one segment). The lengths should be just the
-    wordpiece lengths, not including the SEP and CLS tokens.
-
-    According to the HF glue_utils.py script, the convention is:
-
-    (a) For sequence pairs:
-        tokens: [CLS] is this jack ##son ##ville ? [SEP] no it is not . [SEP]
-        type_ids:   0   0  0    0    0     0     0   0   1  1  1  1   1   1
-
-    (b) For single sequences:
-        tokens:   [CLS] the dog is hairy . [SEP]
-        type_ids:   0   0   0   0  0     0   0
-    """
-    if length2:
-        return [0] * length1 + [0] + [0] + [1] * length2 + [1]
-    else:
-        [0] * length1 + [0] + [0]
-
-
-def get_gpt2_segment_ids(length1: int, length2: int) -> List[int]:
-    """Get an array of segment IDs in GPT2's format, for an input with one or
-    two segments (set length2=0 for one segment). The lengths should be just the
-    wordpiece lengths, not including the SEP and CLS tokens.
-
-    I'm really not sure how this should look? We currently require segment
-    boundaries, so we're just using the <|endoftext|> markers in their vocab?
+    @property
+    def width(self) -> int:
+        for tensor in reversed(self.tensors):
+            if len(tensor.shape) == 3:
+                return tensor.shape[-1]
+        else:
+            raise ValueError("Cannot find last hidden layer")
+
+    def align_to_tokens(self, ops, wp: Floats2d) -> Tuple[Ragged, Callable]:
+        aligned = Ragged(wp[self.trf2tok.data], self.trf2tok.lengths)
+
+        def backprop_tok_alignment(d_aligned: Ragged) -> Floats3d:
+            d_wp = ops.alloc2f(len(d_aligned), d_aligned.data.shape[1])
+            ops.scatter_add(d_wp, self.trf2tok.data, d_aligned.data)
+            return d_wp
```

Comment:
There's a bug here, but I'm not familiar enough with the code (yet) to fix it. The `scatter_add` function is tripping over indices/dimensions. Should `d_wp` be 3D?